subprojects/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/UpgradePropertyInstrumentationExtension.java

```diff
@@ -0,0 +1,250 @@
+import org.gradle.internal.instrumentation.processor.modelreader.api.CallInterceptionRequestReader.Result.Success;
+import org.gradle.internal.instrumentation.processor.modelreader.impl.AnnotationUtils;
+import org.jetbrains.annotations.NotNull;
+import org.objectweb.asm.Type;
+
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.AnnotationValue;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.Modifier;
+import javax.lang.model.type.TypeMirror;
+import java.lang.annotation.Annotation;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+import static org.gradle.internal.instrumentation.model.ParameterKindInfo.METHOD_PARAMETER;
+import static org.gradle.internal.instrumentation.processor.codegen.TypeUtils.*;
+import static org.gradle.internal.instrumentation.processor.modelreader.impl.TypeUtils.extractType;
+
+public class UpgradePropertyInstrumentationExtension
+    extends RequestGroupingInstrumentationClassGenerator
+    implements AnnotatedMethodReaderExtension, ClassLevelAnnotationsContributor, CodeGeneratorContributor {
+
+    private static final Type PROPERTY = Type.getType(Property.class);
+    private static final Type REGULAR_FILE_PROPERTY = Type.getType(RegularFileProperty.class);
+    private static final Type DIRECTORY_PROPERTY = Type.getType(DirectoryProperty.class);
+    private static final Type DEFAULT_TYPE = Type.getType(UpgradedProperty.SameAsGenericType.class);
+    private static final String INTERCEPTOR_DECLARATION_CLASS_NAME = "org.gradle.internal.classpath.InterceptorDeclaration_JvmBytecodeImplCodeQuality";
+
+    @Override
+    public InstrumentationCodeGenerator contributeCodeGenerator() {
+        return this;
+    }
+
+    @Override
+    protected String classNameForRequest(CallInterceptionRequest request) {
+        return request.getRequestExtras().getByType(UpgradePropertyImplementationClass.class)
+            .map(UpgradePropertyImplementationClass::getImplementationClassName)
+            .orElse(null);
+    }
+
+    @Override
+    protected Consumer<TypeSpec.Builder> classContentForClass(
+        String className,
+        Collection<CallInterceptionRequest> requestsClassGroup,
+        Consumer<? super CallInterceptionRequest> onProcessedRequest,
+        Consumer<? super GenerationResult.HasFailures.FailureInfo> onFailure
+    ) {
+        List<MethodSpec> methods = requestsClassGroup.stream()
+            .map(UpgradePropertyInstrumentationExtension::mapToMethodSpec)
+            .collect(Collectors.toList());
+        return builder -> builder.addModifiers(Modifier.PUBLIC).addMethods(methods);
```

Comment:
Please also call `onProcessedRequest` for each request that is successfully processed. The main purpose of this callback is to detect requests that were ignored by all code generators.

Reply:
It does not seem to be an obvious contract, though. Maybe we could handle this in `RequestGroupingInstrumentationClassGenerator` instead, so that all requests that get into a group for some class name are automatically included into the success result, as long as no failure is reported for them.

Reply:
Yes, it's OK to keep the calling code as is for now.

> Maybe `RequestGroupingInstrumentationClassGenerator` could also check if all requests have been either processed or failed. And if not it should throw an error.

I'd even say that a better place for this check would be `InstrumentationCodeGeneratorHost#generateCodeForRequestedInterceptors` (it should consider `InstrumentationCodeGenerator.GenerationResult.CanGenerateClasses#getCoveredRequests`). Yes, I think we should add such a check in the future.